src/core/ddsi/src/ddsi_serdata_default.c

```diff
@@ -656,6 +656,24 @@ static size_t serdata_default_print_raw (const struct ddsi_sertopic *sertopic_co
   return (size_t) snprintf (buf, size, "(blob)");
 }
 
+static void serdata_default_add_keyhash (const struct ddsi_serdata *serdata_common, char *buf, bool force_md5)
+{
+  const struct ddsi_serdata_default *d = (const struct ddsi_serdata_default *)serdata_common;
+  assert(buf);
+  assert(d->keyhash.m_set);
+  if (force_md5 && d->keyhash.m_iskey /* m_iskey == !md5 */)
+  {
+    ddsrt_md5_state_t md5st;
+    ddsrt_md5_init  (&md5st);
+    ddsrt_md5_append(&md5st, (ddsrt_md5_byte_t*)(d->keyhash.m_hash), 16);
```

Comment:
From the DDS Security 1.1 spec (page 42): "The KeyHash shall be computed as the 128 bit MD5 Digest (IETF RFC 1321) applied to the CDR Big-Endian encapsulation of all the Key fields in sequence. Unlike the rule stated in sub clause 9.6.3.3 of the DDS specification, the MD5 hash shall be used regardless of the maximum-size of the serialized key."

I read that as calculating the MD5 over the bytes actually used for representing the key value, not simply all 16 bytes. That information currently gets discarded along the way ... It looks like it would be pretty trivial to let ``dds_stream_extract_keyhash`` and ``dds_stream_write_keyBE`` return the number of bytes: they have available internally anyway as a position in the output buffer and they currently don't return a value anyway.

As to storing it ... it's guaranteed to be ≤ 16 bytes; I'd just use an additional 5 bits large field in ``struct dds_keyhash``.

Reply:
I don't really have an counter argument. So, I think your interpretation is correct.